### ISSUE TITLE
Support for Java 11

### ISSUE THREAD

**Issue by TyrantFox** [cjmx/cjmx#30]
So sorry if this is already addressed somewhere. I've searched everywhere I can think and don't see mention of it. I love cjmx and we use it all over. We're in the midst of upgrading to Java 11 (so we can use HTTP/2).

However in Java 11 the tools.jar is removed from the JAVA_HOME/lib directory and doesn't appear to have a replacement anywhere.

Is there any way to make cjmx work with Java 11?

**Comment by mpilquist** [comment]
@TyrantFox Nice! I didn't know folks were using cjmx. I'm not certain but I think that despite tools.jar being gone, the classes are still available on the classpath. Have you tried adjusting the startup script to avoid requiring tools.jar on the classpath?

I can look in to how we can support Java 11+ soon.

**Comment by TyrantFox** [comment]
Wow! That's a quick response. I'm rolling up my sleeves now to see if I can address it. The error I get is
`
> connect 6864
[error] com.sun.tools.attach.AgentLoadException = Agent JAR not found or no Agent-Class attribute
`
when I don't have the tools.jar.

**Comment by TyrantFox** [comment]
Looks like it's in Attach:41. This makes assumptions about where the jar should be.

`val agent = vm.getSystemProperties.getProperty("java.home") + File.separator + "lib" + File.separator + "management-agent.jar"`

I'm not familiar with Scala, but will troubleshoot in Java.

**Comment by mpilquist** [comment]
Oh yeah, that'll do it! 😬

**Comment by TyrantFox** [comment]
Ref https://bugs.openjdk.java.net/browse/JDK-8179063

The management-agent.jar is removed and two new methods are added to the VirtualMachine. The one I believe we care about is:

`vm.startLocalManagementAgent()`

Not sure what's the good way to make it work for JDK 8 as well. I'll see if I can figure out how to compile this in Scala. If I can make it work I'll share a patch.

**Comment by mpilquist** [comment]
OK cool -- we'll probably need to do some dynamic loading trickery -- e.g. use Java reflection to see if that method exists and if so call it, otherwise fallback to loading management agent.

**Comment by TyrantFox** [comment]
As it turns out, the new method, `startLocalManagementAgent()` is introduced as of Java 8. Here's the diff I applied:

```
diff --git a/src/main/scala/cjmx/util/jmx/Attach.scala b/src/main/scala/cjmx/util/jmx/Attach.scala
index 013a8fb..cbbd6ec 100644
--- a/src/main/scala/cjmx/util/jmx/Attach.scala
+++ b/src/main/scala/cjmx/util/jmx/Attach.scala
@@ -38,8 +38,7 @@ object Attach {
   def localConnectorAddress(vm: VirtualMachine): Either[String, String] = {
     def getLocalConnectorAddress = Option(vm.getAgentProperties.getProperty("com.sun.management.jmxremote.localConnectorAddress"))
     val localConnectorAddress = getLocalConnectorAddress orElse {
-      val agent = vm.getSystemProperties.getProperty("java.home") + File.separator + "lib" + File.separator + "management-agent.jar"
-      vm.loadAgent(agent)
+      vm.startLocalManagementAgent()
       getLocalConnectorAddress
     }
     localConnectorAddress.toRight("Failed to connect to VM ID %s.".format(vm.id))
```


I compiled under Java 8 and it now works both under Java 8 and Java 11 with no tools.jar on the classpath.

**Comment by TyrantFox** [comment]
Would really appreciate if you can point me in the right direction to produce the equivalent -app jar file that you have available for download. This is the first time I'm using sbt or Scala.

**Comment by mpilquist** [comment]
@TyrantFox Nice find, do you want to open a PR with that change? I think requiring Java 8 for a new CJMX release would be totally fine.

To build the -app jar, run `sbt publishLocal`. After that finishes, the app jar should be in `~/.ivy2/local/com.github.cjmx/cjmx_2.12/2.7.1-SNAPSHOT/jars/cjmx_2.12-app.jar`

**Comment by TyrantFox** [comment]
@mpilquist Thank you for your assistance. Really appreciate it.

**Comment by mpilquist** [comment]
@TyrantFox Thanks for all your help on this one. I published a 2.8.0 release. The README links to the app jar.
